File: blockchain/02-OutputsInputsTransactions/src/main/java/no/kalli/DSAUtil.java
package no.kalli;

import javax.xml.bind.DatatypeConverter;
import java.security.*;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;

public class DSAUtil {

    public static KeyPair generateRandomDSAKeyPair() {
        try {
            KeyPairGenerator keyGen = KeyPairGenerator.getInstance("DSA");
            keyGen.initialize(512);
            return keyGen.generateKeyPair();

        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static byte[] signWithDSA(PrivateKey privateKey, String message) {

        try {
            Signature dsa = Signature.getInstance("DSA");
            dsa.initSign(privateKey);
            dsa.update(message.getBytes());
            return dsa.sign();

        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static boolean verifyWithDSA(PublicKey publicKey, String message, byte[] signature) {
        try {
            Signature dsa = Signature.getInstance("DSA");
            dsa.initVerify(publicKey);
            dsa.update(message.getBytes());
            return dsa.verify(signature);

        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static String base64EncodeKey(Key key) {
        return DatatypeConverter.printBase64Binary(key.getEncoded());
    }

    public static PublicKey base64DecodePublicKey(String base64Key) {
        byte[] bytes = DatatypeConverter.parseBase64Binary(base64Key);
        try {
            return KeyFactory.getInstance("DSA")
                    .generatePublic(new X509EncodedKeySpec(bytes));

        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    public static PrivateKey base64DecodePrivateKey(String base64Key) {
        byte[] bytes = DatatypeConverter.parseBase64Binary(base64Key);
        try {
            return KeyFactory.getInstance("DSA")
                    .generatePrivate(new PKCS8EncodedKeySpec(bytes));

        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }


}

File: blockchain/02-OutputsInputsTransactions/src/main/kotlin/no/kalli/DSAUtil.kt
package no.kalli

import java.security.*
import java.security.spec.InvalidKeySpecException
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import javax.xml.bind.DatatypeConverter

object DSAUtil {

    fun generateRandomDSAKeyPair(): KeyPair {
        try {
            val keyGen = KeyPairGenerator.getInstance("DSA")
            keyGen.initialize(512)
            return keyGen.generateKeyPair()

        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    }

    fun signWithDSA(privateKey: PrivateKey, message: String): ByteArray {

        try {
            val dsa = Signature.getInstance("DSA")
            dsa.initSign(privateKey)
            dsa.update(message.toByteArray())
            return dsa.sign()

        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    }

    fun verifyWithDSA(publicKey: PublicKey, message: String, signature: ByteArray): Boolean {
        try {
            val dsa = Signature.getInstance("DSA")
            dsa.initVerify(publicKey)
            dsa.update(message.toByteArray())
            return dsa.verify(signature)

        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    }

    fun base64EncodeKey(key: Key): String {
        return DatatypeConverter.printBase64Binary(key.encoded)
    }

    fun base64DecodePublicKey(base64Key: String): PublicKey {
        val bytes = DatatypeConverter.parseBase64Binary(base64Key)
        try {
            return KeyFactory.getInstance("DSA")
                    .generatePublic(X509EncodedKeySpec(bytes))

        } catch (e: InvalidKeySpecException) {
            throw RuntimeException(e)
        } catch (e: NoSuchAlgorithmException) {
            throw RuntimeException(e)
        }

    }

    fun base64DecodePrivateKey(base64Key: String): PrivateKey {
        val bytes = DatatypeConverter.parseBase64Binary(base64Key)
        try {
            return KeyFactory.getInstance("DSA")
                    .generatePrivate(PKCS8EncodedKeySpec(bytes))

        } catch (e: InvalidKeySpecException) {
            throw RuntimeException(e)
        } catch (e: NoSuchAlgorithmException) {
            throw RuntimeException(e)
        }

    }


}
